contracts/BSCValidatorSet.sol

```diff
@@ -479,7 +501,178 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     return v1.consensusAddress == v2.consensusAddress && v1.feeAddress == v2.feeAddress && v1.BBCFeeAddress == v2.BBCFeeAddress && v1.votingPower == v2.votingPower;
   }
 
+  function _misdemeanor(address validator) private returns (uint256) {
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
+      return index;
+    }
+    // the actually index
+    index = index - 1;
+
+    uint256 income = currentValidatorSet[index].incoming;
+    currentValidatorSet[index].incoming = 0;
+    uint256 rest = currentValidatorSet.length - 1;
+    emit validatorMisdemeanor(validator,income);
+    if (rest==0) {
+      // should not happen, but still protect
+      return index;
+    }
+    uint256 averageDistribute = income/rest;
+    if (averageDistribute!=0) {
+      for (uint i=0;i<index;i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+      uint n = currentValidatorSet.length;
+      for (uint i=index+1;i<n;i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    return index;
+  }
+
+  function _felony(address validator) private {
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
```

Comment:
if index == 0 or index == 1, it will both return 0

Reply:
fixed